src/ngraph/placement.cpp

```diff
@@ -224,3 +224,186 @@ pair<vector<shared_ptr<Function>>, unordered_map<shared_ptr<op::Parameter>, shar
+    for (auto cluster : clusters)
+    {
+        for (auto node : cluster)
+        {
+            if (cluster_nodes.find(node) != cluster_nodes.end())
+            {
+                throw ngraph_error("Node " + node->get_name() + " is duplicated in clusters");
+            }
+            cluster_nodes.insert(node);
+        }
+    }
+    unordered_set<shared_ptr<Node>> f_nodes;
+    for (auto node : f->get_ordered_ops())
+    {
+        f_nodes.insert(node);
+    }
+    if (cluster_nodes != f_nodes)
+    {
+        throw ngraph_error(
+            "Cluster's nodes are not the same as function's nodes. cluster_nodes.size()=" +
+            to_string(cluster_nodes.size()) + ", f_nodes.size()=" + to_string(f_nodes.size()));
+    }
+
+    return clusters;
+}
+
+// Suffix *_size  as a part of function name is temporary, this suffix
+//  will be removed when the backends move to the latest Hybrid backend
+pair<vector<shared_ptr<Function>>, unordered_map<shared_ptr<op::Parameter>, shared_ptr<op::Result>>>
+    ngraph::split_function_by_placement_size(const shared_ptr<Function>& f)
+{
+    // Split functions to clusters of nodes that can be computed together
+    vector<unordered_set<shared_ptr<Node>>> clusters = group_function_nodes_to_clusters_size(f);
+
+    // Map from (intermediate) parameter to result node, for guiding data copy among devices
+    unordered_map<shared_ptr<op::Parameter>, shared_ptr<op::Result>> map_parameter_to_result;
+
+    // Split neighboring nodes if they belong to different clusters
+    // TODO: optimization to group multiple result node from the same source,
+    //       and to group the parameter node in the same cluster with the same result node source
+    unordered_map<shared_ptr<Node>, unordered_set<shared_ptr<Node>>*> map_node_to_cluster;
+    for (auto& cluster : clusters)
+    {
+        for (auto node : cluster)
+        {
+            map_node_to_cluster[node] = &cluster;
+        }
+    }
+    for (auto dst_node : f->get_ordered_ops())
+    {
+        for (auto src_node : dst_node->get_arguments())
+        {
+            auto src_cluster = map_node_to_cluster.at(src_node);
+            auto dst_cluster = map_node_to_cluster.at(dst_node);
+            if (src_cluster != dst_cluster)
+            {
+                // Split src_node and dst_node
+                pair<shared_ptr<op::Result>, shared_ptr<op::Parameter>> res_par_pair =
+                    insert_result_parameter_split_size(src_node, dst_node);
+                shared_ptr<op::Result> res_node = res_par_pair.first;
+                shared_ptr<op::Parameter> par_node = res_par_pair.second;
+                map_parameter_to_result[par_node] = res_node;
+
+                // Insert newly created nodes into clusters
+                src_cluster->insert(res_node);
+                dst_cluster->insert(par_node);
+            }
+        }
+    }
+
+    // Create functions from clusters
+    vector<shared_ptr<Function>> sub_functions;
```

Comment:
nitpick: this could've been a standalone helper.

Reply:
find this easier, shall I can make it a lamda so it can be serve both purposes readability as well as same function ?